lib/broadway/producer.ex

```diff
@@ -400,7 +400,7 @@ defmodule Broadway.Producer do
   defp enqueue_batch_r(queue, list), do: :queue.in_r(list, queue)
 
   defp rate_limit_messages(_state, [], _count) do
-    {[], []}
+    {[], [], []}
```

Comment:
This will break. The first element of the tuple needs to be the rate limiting state. If we're processing no messages, we should leave it as what it was before, which is `state.rate_limiting.state`. You should change this clause to this:

```elixir
defp rate_limit_messages(state, _table_name, [], _count) do
  {state.rate_limiting.state, [], []}
end
```

and add `_state` as the first argument to the next clause.

Reply:
We also need a test. :)

Reply:
Yeah this is a bit nasty to test I'm afraid, because to get here you'd have to have a non-empty buffer but then change the rate limiting between the moment we checked the buffer and here so that it closes and we say that we can't send any messages. :|

Reply:
Not even :sys.suspend can happen with that? Or have a message in the buffer that takes long to process?

Reply:
@josevalim the race condition that we would have to trigger happens between two function calls basically, not between processing of messages, so here it feels a bit not worth it for the investment to test this, but up to you :)

Reply:
(also I wouldn't know how to trigger this, might have to have a better look)

Reply:
Good call on the `state.rate_limiting.state`. I'll make that change right away (although I could have sworn that once I made the change my errors went away and I was able to get a consistent processing rate...i'll verify with a repo that reproduces the error consistently).

Reply:
@amokan if you have a repo that reproduces errors, do you think you can come up with a test that reproduces this?

Reply:
^ Guessing you meant to tag me :P ? I'll try and get something put together tonight and add it to the PR.

Reply:
@whatyouhide I'm still trying to wrap my head around how all the rate limiting stuff works and how best to test it (it seems like it is a token bucket style rate limiter ?).

Changing the first item in the tuple to `state.rate_limiting.state` won't actually work since the first param to the `rate_limit_messages/3` function is the ETS table name and not the state (https://github.com/plataformatec/broadway/blob/8729382b510c32072f6c6545e408fb887bf0c481/lib/broadway/producer.ex#L341-L345). Probably just a lingering variable name that isn't correct.

I was able to create a repo to reproduce this very consistently (seems rather deterministic), so it may be easier to test than anticipated as it may not be a timing related thing: https://github.com/akoutmos/broadway_error_repro

Lastly, the return of `{[], [], []}` that I had initially seems to work (and actually anything as the first element seems to work fine including `open` and `closed`) because the RateLimiter does a `send(pid, {__MODULE__, :reset_rate_limiting})` and that resets the state in `Broadway.Producer` (at least that's what it looks like).

Sorry for the long winded comment...hopefully the provided information helps :).